### ISSUE TITLE
Review Output Selection policy for wallet transfer

### ISSUE THREAD

**Issue by SebastienGllmt** [Emurgo/yoroi-frontend#1067]
Currently transferring to Yoroi through the "transfer" feature sends all the Ada to the next unused internal address. This is at least some privacy (will never be reused) but arguably not ideal. The Cardano-Wallet team has a different implementation here: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Migration.hs

**Comment by KtorZ** [comment]
This implementation came out of the original analysis here under. Of all highlighted solutions, we implemented the 3rd. 

- We want users to be able to migrate their funds from a legacy random wallet to a new sequential wallet. To do this, we have to move funds from a wallet to another by making transactions. Funds are ultimately a sum of many coins (a.k.a UTxOs). In a transaction, we can select a few coins, and send them to addresses, effectively creating new coins / UTxOs doing this.

- There are some limitations regarding the number of coins that can be selected at once in a single transaction (theoretically 255 coins, in practice ~170) because there's a transaction max size (in bytes) enforced by the network. Also, there's a direct relationship between the maximum number of inputs we can select, and the maximum number of outputs we can produce (increasing one will decrease the other, and vice-versa).

- When making a transaction, coins used as inputs for a transaction becomes unavailable for a while, until the transaction is inserted into the ledger and, make some new coins available as change (very much like when paying with bank notes to a shop, if we give a 20 EUR note to pay for 3 EUR, we can't spend the remaining 17 EUR before we have received the change!). So, a wallet with a small number of UTxO will not be able to make many transactions in parallel and will have to make them sequentially, waiting for the previous ones to be inserted before making new ones (we also say that a wallet is not "fragmented enough").

- We can't have more than a fixed number of unused addresses in a sequential wallet. By default this number is `20` but can go up to `100` (so-called "address_pool_gap" in the API). This means that, even if we wanted to, we couldn't make a transaction to a new wallet that uses more than `address_pool_gap` different outputs.

So, here's the problem to solve:
For many users, we will likely have to make several transactions. It's quite common to see wallets with a thousand UTxOs or more, so this would require at least 8 or 9 transactions (assuming the current soft limit for transaction max size). The way we make transactions and construct outputs will shape the available UTxOs in the new wallet. We have different approaches to this (non-exhaustive):

1. (straightforward) We make only transactions with single outputs, merging as many inputs as we can into a single UTxOs.

> **Drawbacks** 
> 
> we are effectively reducing the possible "parallelism" of the new wallet which could have a very dramatic impact on the wallet usability. Less UTxOs, means less concurrent transactions.

2. (expensive) We try to make a a 1:1 mapping, and produce a new output for every input, trying to re-create a wallet that is as close as possible to the source one.

> **Drawbacks**
> 
> This is probably quite expensive for big wallets, especially if they contain a lot of small coins (a.k.a dust). we'll be filling outputs with many many small coins and they will take space and cost money to transfer.

3. (expensive-smart) We fix a threshold for dust, making sure that any coin that is less than a given small value is merged into a single coin. This way, we are effectively cleaning up small coins.

> **Drawback**
>
> - If a user has a wallet with only dust, then we'll end up with a resulting wallet that has only a single big UTxO. Not cool.

4. (lazy) We let the API client chose the number of UTxOs they want in the new wallet, and normalize all UTxOs to be of the same size, with a number equal to what the client specified.

> **Drawbacks**
>
> - Could require some extra work for API clients if they want to expose this feature to users. They could still fix it to an arbitrary value themselves and not ask the users.

5. (smarter) Look at the past transaction history for that user and the average & median transaction amount and generate outputs that are roughly of this average size. This will so-to-speak optimize a user wallet based on its own usage.

> **Drawbacks**
> 
> - This is somewhat more complicated for an action that will occur once in the lifetime of a wallet
> - It is unsure that this will have better results than other, less complicated approach

6. (lazy2) Pick an arbitrary new size that is "reasonable" and migrate funds to addresses, evenly distributed 

> **Drawbacks**
>
> - We are still constrained by the fact that we can't send to more than `address_pool_gap` addresses.
> - We might create wallets that are unapdated to a particular user

**Comment by jarstorm** [comment]
Hello, and thank you for your contribution.
We're currently reviewing long-standing issues to keep the tracker manageable. As this issue hasn't seen activity in some time, we're closing it.
If the problem persists or you'd like to follow up, feel free to comment and we'll reopen it.
